opm/simulators/wells/WellAndGroupStates.hpp

```diff
@@ -0,0 +1,564 @@
+    std::vector<std::string> well_names_;
+    std::map<std::string, GroupState> group_states_;
+
+    std::vector<std::vector<PerforationData>> well_perf_data_;
+
+
+    // -----------  Private functions  -----------
+
+    template <unsigned long N>
+    static double sum(const std::array<double, N>& x)
+    {
+        return std::accumulate(x.begin(), x.end(), 0.0);
+    }
+
+    static void initSingleWell(const std::vector<double>& cell_pressures,
+                               const Schedule& schedule,
+                               const Well& well,
+                               const int report_step,
+                               const PhaseUsage& phase_usage,
+                               const std::vector<PerforationData>& perf_data,
+                               const SummaryState& summary_state,
+                               SingleWellState<NumActivePhases>& wstate)
+    {
+        const auto inj_controls = well.isInjector() ? well.injectionControls(summary_state) : Well::InjectionControls(0);
+        const auto prod_controls = well.isProducer() ? well.productionControls(summary_state) : Well::ProductionControls(0);
+
+        // Initialize flag, statuses and current controls.
+        initFlagsAndControls(schedule, well, report_step, inj_controls, prod_controls, wstate);
+
+        // Initialize bhp, thp and well rates.
+        initPressuresAndRates(cell_pressures, well, phase_usage, perf_data, inj_controls, prod_controls, wstate);
+
+        // Initialize connection pressures and rates.
+        initConnections(cell_pressures, perf_data, wstate);
+
+        // Initialize multi-segment well parts (the 'segments' member).
+        if (well.isMultiSegment()) {
+            initMultiSegment(well, wstate);
+        }
+    }
+
+
+
+
+    static void initSingleWellFromPrevious(const SingleWellState<NumActivePhases>& prev_wstate,
+                                           SingleWellState<NumActivePhases>& wstate)
+    {
+        // Flags and statuses.
+        wstate.status = prev_wstate.status;
+        wstate.is_producer = prev_wstate.is_producer;
+        wstate.effective_events_occurred = prev_wstate.effective_events_occurred;
+        wstate.current_injection_control = prev_wstate.current_injection_control;
+        wstate.current_production_control = prev_wstate.current_production_control;
+
+        // Quantities.
+        wstate.bhp = prev_wstate.bhp;
+        wstate.thp = prev_wstate.thp;
+        wstate.temperature = prev_wstate.temperature; // 20 degrees Celcius by default.
+        wstate.surface_rates = prev_wstate.surface_rates;
+        wstate.reservoir_rates = prev_wstate.reservoir_rates;
+        wstate.dissolved_gas_rate = prev_wstate.dissolved_gas_rate;
+        wstate.vaporized_oil_rate = prev_wstate.vaporized_oil_rate;
+        wstate.potentials = prev_wstate.potentials;
+
+        // Connections.
+        if (wstate.connections.size() == prev_wstate.connections.size()) {
```

Comment:
What if this tests fails - i.e. a new connection is added, then *all* the connections will be uninitialized. Would it make sense to initialize the common connections?

Reply:
I think so, but my initial goal is to refactor this with *not a bit changed* in the results, so that the refactoring can be considered as a separate thing. Improving behaviour will therefore come only later.